### DISCUSSION TITLE
Can I obtain the code length using the MIR API or any specific variable?

### DISCUSSION THREAD

**Discussion by qwe661234** [vnmakarov/mir#338] [Q&A]
Currently, I am aware that the `MIR_gen` function retrieves the `code_len` from the `target_translate` function. However, I'm interested in obtaining the `code_len` after the `c2mir_compile` finishes. Is it possible to use the MIR API or any specific variables to retrieve the `code_len` at that stage?

If not, how can I create an API to accomplish this task with minimal changes?

**Comment by vnmakarov** [comment]
`c2mir_compile` only produces MIR-code without any machine code.  But there is no API to get machine code length even by MIR API.  The function machine code length appears only on debugging output of MIR-generator.

I am just curious what you are trying to get:
  * length of all generated code including machine code of functions, bss, data etc:
  * all allocated memory for machine code of all functions (it is done by pages and some pages can contain unused memory)
  * or machine code length allocated for each function
 
I think, the simplest implementation would be providing machine code length for each function.  Then after using c2mir_compile and mir-generator it is possible to traverse all functions in all modules to get overall length.

So what is your thoughts about this?

**Reply by qwe661234** [reply]
I try to get is machine code length for each function, because I want to copy machine code to a specific memory space.

**Comment by vnmakarov** [comment]
> I try to get is machine code length for each function, because I want to copy machine code to a specific memory space.

Generally speaking, I can add a member to mir func struct about the code length.

I only can not understand why do you need to move the code. The code is not position independent.  It can not be moved somewhere else and executed at the new place.

**Reply by qwe661234** [reply]
We aim to utilize MIR as a compiler for JIT and copy machine code to code cache for better management. Thank you for your assistance, and please let me know if there are any updates!

**Comment by qwe661234** [comment]
Thanks for your reply. I have refined my question. I want to leverage MIR as a compiler to integrate a JIT compiler into a virtual machine project. Initially, I simply calculate the size of machine code generated by MIR and copy the machine code to the code cache for better memory management. This approach works when the host machine architecture is x86; however, it fails when the host machine architecture is aarch64. Is there a possible solution or any recommendations to solve this issue?

**Comment by vnmakarov** [comment]
> Initially, I simply calculate the size of machine code generated by MIR and copy the machine code to the code cache for better memory management. This approach works when the host machine architecture is x86; however, it fails when the host machine architecture is aarch64. Is there a possible solution or any recommendations to solve this issue?

Again, MIR does not generate position-independent code.  So where the code was generated, only there it can be executed.

Saying that, there is a possibility for some generated code to be position-independent.  x86-64 uses pc-relative addressing in many cases and for such cases it can be position-independent.  If a MIR-code contains switch insn (used for implementing C switch-stmt), it is definitely not position-independent code on any target.
